Various readability improvements to Discus algorithm

DannyHindson opened the pull request:
**Description of work.**

These are changes that I should have put through on #32744 but somehow forgot to push the commit
Add comments to integrateCumulative method
Move some const method parameters into private variables to avoid them tunneling through multiple levels of function call
Note:
The Sample class doesn't provide access to its internal shared pointer to the shape for some reason. So have to clone the shape (which could be large) to get a shared pointer to it

**To test:**

These don't change the functionality at all. They're just code tidies so code review only

Fixes #32208.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Various readability improvements

Add comments to integrateCumulative method
Move some const method parameters into private variables to avoid them
tunneling through multiple levels of function call
Note:
The Sample class doesn't provide access to its internal shared
pointer to the shape for some reason. So have to clone the
shape (which could be large) to get a shared pointer to it

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -197,10 +197,12 @@ void DiscusMultipleScatteringCorrection::exec() {
   }
 
   MatrixWorkspace_sptr sigmaSSWS = getProperty("ScatteringCrossSection");
+  if (sigmaSSWS)
```
If this fails does it make sense to continue onwards?
Should we instead have something like:
```
if (!sigmaSSWS){
    throw std::invalid_argument("ScatteringCrossSection was null");
}
```

Alternatively, does validateInputs should pick this up? In which case an assert should be fine here as this shouldn't fail unless a future developer changes validateInputs

DannyHindson replied:
This is meant to be an optional parameter so I think this is OK. If you look at line 497 there's code to deal with this not being supplied.

```
  if (m_sigmaSS) {
    scatteringXSection = interpolateFlat(m_sigmaSS, kinc);
  } else {
    scatteringXSection = material.totalScatterXSection();
  }
```

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -235,13 +237,13 @@ void DiscusMultipleScatteringCorrection::exec() {
                                           : static_cast<int64_t>(inputWS->getNumberHistograms());
 
-  const auto &sample = inputWS->sample();
+  m_sampleShape = std::shared_ptr<const Geometry::IObject>(inputWS->sample().getShape().clone());
```
Is there a way to simply get the shared_ptr without cloning? That way we hold the lifetime too but don't have to do a potentially expensive copy

DavidFair replied:
Equally, if we hold the sptr for inputWs will that indirectly keep the lifetime for the underlying shape?
We might be able to hold onto a ref and rely on the fact we've pinned the `inputWs` to manage lifetime :thinking:

DannyHindson replied:
There's another nasty looking clone() in the MonteCarloAbsorption algorithm (in MCInteractionVolume.cpp) which is probably someone else trying to work around the same problem. I think I'll just add an extra getShapePtr method onto the Sample class so I can (separately) fix that up as well

DavidFair commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -534,17 +534,17 @@ double DiscusMultipleScatteringCorrection::interpolateSquareRoot(const Histogram
-    return histToInterpolate.y().back();
+double DiscusMultipleScatteringCorrection::interpolateFlat(
+    std::shared_ptr<const Mantid::HistogramData::Histogram> histToInterpolate, double x) {
+  if (x > histToInterpolate->x().back()) {
```
There's lots of derefs through all of this. Can we simply take a ref at the start to cut the noise and number of derefs the compiler has to do down?
```
auto &xHisto = histToInterpolate->x()
auto &yHisto = histToInterpolate->y()

if (x < xHisto.back())
// ...etc.
```

DavidFair reviewed (changes requested):
Just some minor nits / questions
Overall looks a lot better and makes it clear which params we expect to change per-iteration vs those which change once per alg execution

DannyHindson committed:
Stop cloning sample shape

Add new method to API::Sample that provides access to the m_shape
shared ptr. The m_shape variable was originally an instance rather
than a pointer and I think the getShape() method was never updated

Also tidy up some dereferencing to make code easier for reader and
compiler

DavidFair reviewed (approved):
LGTM, thanks for doing those changes :+1:
